Dog wit the butter (feat. a bunch of dog-related code improvements)

Sealed101 opened the pull request:
## About The Pull Request
Adds a `dog_fashion` for the stick of butter.(screenshot is outdated as Lisa won't have butter no more)
![butter dawgs](https://github.com/tgstation/tgstation/assets/75863639/a22e702c-98a8-4283-abd9-28d4a9fb3bd0)

Also cleans up dog.dm because it was SHIT and FUCK and MY FUCKING GOD TWO INITIALIZE()s TWO TIMES IN A SINGLE FILE WHAT IN THE GODDAMN

Most noticeably, Lisa properly won't wear any hats, and puppies properly can't wear head/back items (by just removing those item slots from the strip/equip menu. if some admeme wants to fumble around they may still equip shit there. but otherwise for a normal player those slots are inaccessible).
Basic mobs now also send signals when they run `appear_dead`/`appear_alive` procs, which corgis hook into to update their dead fashion overlays.
The side-effect of getting that to work is that dogs (and any basic mob that uses `play_dead` ai behavior) are so good at feigning death, that they fool medical HUDs and other related things. They're just that good.
There's a bunch of other things involved and I was mostly just being angry at the state of the file so I'll check back when I gather all things changed.

![strippy](https://github.com/tgstation/tgstation/assets/75863639/ec4d17a2-d4df-401c-bd1f-7c4ee1b95671)


## Why It's Good For The Game

https://github.com/tgstation/tgstation/assets/75863639/b34589cb-94d6-4b80-bf0f-1814c08da100



## Changelog
:cl:
add: dog with a butter on 'em
add: dead dog with da butter on 'em (dogs feigning death are so good at it, they appear dead to medical HUDs and other things)
add: Nars-Ian now can revive from the dead if he consumes a pet
fix: fixes dog fashion items with no speech modifiers set making dressed up corgis unable to perform their speech or emote behaviors
fix: fixes old Ian losing his mobility ride when shaved with a razor
fix: fixes pets not dropping their collar when gibbed
fix: butter don't go on Lisa and corgi puppies (Lisa won't wear hats and corgi puppies can't wear hats and back slot items)
/:cl:

Sealed101 committed:
butter dog

NamelessFairy commented:
New speedrun category

distributivgesetz commented:
dawg wit da butter butter dawg

JohnFulpWillard commented on code/datums/dog_fashion.dm:
```diff
@@ -211,3 +211,15 @@
+	emote_hear = list("licks a trickle of the butter up.")
+
+/datum/dog_fashion/head/butter/New(mob/M)
+	desc = "NAME. [M.real_name] with the butter. NAME. [M.real_name] with a butter on 'em."
```
what does NAME mean

Sealed101 replied:
it gets replaced with the dog's full name, e. g. _Butter Mr. Woofsworth the IV, Esq._
https://github.com/tgstation/tgstation/blob/b488a846a1b469803bbc2cfca7b5621cfd4559ac/code/datums/dog_fashion.dm#L16-L18

Sealed101 replied:
i'm going to change the placeholders so they won't clash with each other and add the REAL_NAME one to the desc as well to use here and in another item

Sealed101 committed:
Update dog_fashion.dm

JohnFulpWillard says: where is the big smile
today, 19:18 MSK, # sprite-club

Sealed101 committed:
swap out the placeholder for name/real_name
so they can't trip over each other

Farquaar commented:
Always nice to have a new use for butter. Maybe in the future we'll see new interactions for butter on a stick.

Wallemations reviewed (approved):
Butter dawg

BlueMemesauce commented:
butterslice exists as an item in game and would fit the meme better, but i guess it would be too hard to see in the sprite

Sealed101 commented:
> butterslice exists as an item in game and would fit the meme better, but i guess it would be too hard to see in the sprite

yeah

Sealed101 committed:
transfer overlay building into update_overlays()
suprisingly the death state works fine
but this doesn't handle play_dead() in any way, need to fix
also dmdoc and shift vars around in dog_fashion.dm

Sealed101 commented:
dog wit da `update_overlays()` on 'im, aka i'm trying to fix things adjacent to the feature because why the hell not

Sealed101 committed:
me when the butter on him
forgot to remove these two from my fumblign attempts at making fake_death work heeheehoohoo

Sealed101 committed:
makes fake_death() update overlays
which means fashion items properly display on a fake dead corgi
however, corgis(and anything else that may use the play_dead ai behavior) are now so good at feigning death that they will fool health HUDs and probably something else (TRAIT_FAKEDEATH). i see this as an absolute win
basic mobs also send signals when they execute look_dead() or look_alive()

MrMelbert commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -227,7 +269,7 @@
 /mob/living/basic/pet/dog/corgi/Initialize(mapload)
 	. = ..()
-	regenerate_icons()
+	update_icon()
```
```suggestion
	update_appearance()
```

Sealed101 replied:
https://github.com/tgstation/tgstation/pull/77039/commits/a8845467d22a084158204c444dc2f8055042edd8

MrMelbert commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -247,7 +289,7 @@
 /mob/living/basic/pet/dog/corgi/death(gibbed)
 	..(gibbed)
-	regenerate_icons()
+	update_icon()
```
```suggestion
	update_appearance()
```

MrMelbert commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -281,7 +323,7 @@ GLOBAL_LIST_INIT(strippable_corgi_items, create_strippable_list(list(
 	corgi_source.inventory_head = null
 	corgi_source.update_corgi_fluff()
-	corgi_source.regenerate_icons()
+	corgi_source.update_icon(UPDATE_OVERLAYS)
```
```suggestion
	corgi_source.update_appearance(UPDATE_OVERLAYS)
```

MrMelbert commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -353,7 +395,7 @@ GLOBAL_LIST_INIT(strippable_corgi_items, create_strippable_list(list(
 	corgi_source.inventory_back = equipping
 	corgi_source.update_corgi_fluff()
-	corgi_source.regenerate_icons()
+	corgi_source.update_icon(UPDATE_OVERLAYS)
```
```suggestion
	corgi_source.update_appearance(UPDATE_OVERLAYS)
```

MrMelbert commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -363,7 +405,7 @@ GLOBAL_LIST_INIT(strippable_corgi_items, create_strippable_list(list(
 	corgi_source.inventory_back = null
 	corgi_source.update_corgi_fluff()
-	corgi_source.regenerate_icons()
+	corgi_source.update_icon(UPDATE_OVERLAYS)
```
```suggestion
	corgi_source.update_appearance(UPDATE_OVERLAYS)
```

MrMelbert commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -402,7 +444,7 @@ GLOBAL_LIST_INIT(strippable_corgi_items, create_strippable_list(list(
 	corgi_source.access_card = null
 	corgi_source.update_corgi_fluff()
-	corgi_source.regenerate_icons()
+	corgi_source.update_icon(UPDATE_OVERLAYS)
```
```suggestion
	corgi_source.update_appearance(UPDATE_OVERLAYS)
```

MrMelbert commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -482,7 +524,7 @@ GLOBAL_LIST_INIT(strippable_corgi_items, create_strippable_list(list(
 		src.inventory_head = item_to_add
 		update_corgi_fluff()
-		regenerate_icons()
+		update_icon(UPDATE_OVERLAYS)
```
```suggestion
		update_appearance(UPDATE_OVERLAYS)
```

Sealed101 committed:
restrict equip slots for puppies and Lisa
(tries to)
need to fiddle with the strippable datums a tad

Sealed101 committed:
but first
REGROUP THE PROCS GOD FUCKING DAMN IT

Sealed101 committed:
brings more sanity to the code
fixes things fumbled in the previous commit
properly sets up Lisa and puppies having different strip menu inventories (Lisa won't wear a hat and puppies can't wear hats and back slot items)
moved Ian's successful survival check and Write_Memory to a roundend event

there's a runtime with empty speak lists that i will look into tomorrow

Sealed101 committed:
further fixes
should fix that hard delete of the callback
fixes speech subtrees receiving null lists on hats with no speech modifiers
more named dogs get their real name set

Sealed101 committed:
fixes old Ian losing his mobility wheels-
when shaved with a razor

Sealed101 committed:
a dead Nars-Ian will revive when consuming a pet
or comes under a narsie_act()
because Nar'Sie cares for her pets unlike someone else

Sealed101 committed:
fixes pets not dropping collars when gibbed

Sealed101 committed:
goddamnit Hoppy

Sealed101 commented:
ight so i think that's all that i found and what i cared to do
- empowers Nars-Ian by letting him revive when consuming other pets (#53517)
- fixes old Ian losing his mobility ride when shaved
- fixes dog fashion items with no set speech modifiers nulling speech for the ai controller, which would result in corgis unable to perform any speech or see/hear emotes
- more named dogs have their names saved in `real_name`, so named corgis won't revert to being `Butter corgi` or something
- fixes pet mobs not dropping their collar when gibbed
- it would appear that `Write_Memory()` was intended to only be called once, either on Ian's survival at the round end or on his death. It could be worth only writing memory upon Ian's complete destruction and round end, so if Ian dies with his body intact and someone revives him, he can still live to the round end and pass his data to the next round. For now, I've just made `memory_saved` var work.
- Lisa and corgi puppies were intended to not have hats equippable, puppies were also intended to not wear back slot items. This was not preserved when the strip menu was reworked, but now it "works", as opening the strip menu for these mobs will just not show the mentioned slots. however, they would still be able to equip hats via `deadchat_control()`. Since Ian can spawn as a puppy if he had died in the previous shift, and there's the Ian's adventure station trait, that is kind of in the realm of possibility. I'm a tad burned out to fix that unfortunately.
- now that I think about it, seems like anything previously mentioned in the PR should account for all of the changes.
- and uh, butter dog, of course 🧈🐕

san7890 commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -619,45 +633,76 @@ GLOBAL_LIST_INIT(strippable_corgi_items, create_strippable_list(list(
+		place_on_head(new saved_head)
 
+///Checks whether Ian has survived the round or not
+/mob/living/basic/pet/dog/corgi/ian/proc/check_Ian_survival()
```
full snake case, I know Ian is a proper noun but we still want it to be `check_ian_survival()`

san7890 commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -619,45 +633,76 @@ GLOBAL_LIST_INIT(strippable_corgi_items, create_strippable_list(list(
 	held_state = "narsian"
 
+//this could maybe be turned into an element for sanity but i don't want to get too distracted
+//i mean i was just placing butter on a corgi yet here i am now
```
```suggestion
```

sorry but code comments should be pertinent. just leave a "TODO make element" or something

san7890 commented on code/modules/mob/living/basic/pets/dog.dm:
```diff
@@ -750,48 +766,58 @@ GLOBAL_LIST_INIT(strippable_corgi_items, create_strippable_list(list(
 
-/mob/living/basic/pet/dog/corgi/puppy/void/Process_Spacemove(movement_dir = 0, continuous_move = FALSE)
-	return 1 //Void puppies can navigate space.
+//Less exciting dog species
```
make a new file for pugs, and also one file for corgi. should be like

```txt
code/modules/mob/basic/pets/dog
| _dog.dm (all root dog code)
| corgis.dm (because it's like 500 lines it's okay to snowflake)
| strippable_items.dm (this could be in _dog.dm, unsure)
| dog_commands.dm (this can be in _dog.dm if you want)
| dog_subtypes.dm (pugs and bullterriers)
```

long files always make my head spin and make these PRs with gargantuan diffs just that harder to review so would be greatly appreciated.

san7890 reviewed (changes requested):
nits, mainly file organization stuff since that would be fantastic to get squared away while you're doing a lot of this reorganization stuff

Sealed101 committed:
splits the doggy litter into separate files
i've left dog commands in the base dog file
also adresses some other review nits and whatnot
